File: tests/fixtures/sample.cs
// sample.cs — fixture file for timebomb scanner tests.
//
// Annotation inventory (hardcoded dates, never relative to today):
//   Expired        (2018–2021): 4
//   Expiring-soon  (2025-06):   1
//   Future / OK    (2088/2099): 2

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;

namespace Sample;

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

public class AppException : Exception
{
    public string Code       { get; }
    public int    StatusCode { get; }
    public object Details    { get; }

    public AppException(string code, string message, int statusCode = 500, object details = null)
        : base(message)
    {
        Code       = code;
        StatusCode = statusCode;
        Details    = details;
    }
}

public class ValidationException : AppException
{
    public ValidationException(string message, object details = null)
        : base("VALIDATION_ERROR", message, 422, details) { }
}

public class NotFoundException : AppException
{
    public NotFoundException(string resource)
        : base("NOT_FOUND", $"{resource} not found", 404) { }
}

public class ConflictException : AppException
{
    public ConflictException(string message)
        : base("CONFLICT", message, 409) { }
}

// ---------------------------------------------------------------------------
// Result type
// ---------------------------------------------------------------------------

public abstract record Result<T>
{
    public record Ok(T Value)    : Result<T>;
    public record Err(string Message) : Result<T>;

    public bool IsOk  => this is Ok;
    public bool IsErr => this is Err;

    public Result<U> Map<U>(Func<T, U> fn) => this switch
    {
        Ok ok   => new Result<U>.Ok(fn(ok.Value)),
        Err err => new Result<U>.Err(err.Message),
        _       => throw new InvalidOperationException()
    };

    public T UnwrapOr(T fallback) => this is Ok ok ? ok.Value : fallback;
}

public static class Result
{
    public static Result<T> Ok<T>(T value)        => new Result<T>.Ok(value);
    public static Result<T> Err<T>(string message) => new Result<T>.Err(message);

    public static async Task<Result<T>> TryAsync<T>(Func<Task<T>> fn)
    {
        try   { return Ok(await fn()); }
        catch (Exception ex) { return Err<T>(ex.Message); }
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

public record ValidationFailure(string Field, string Message, string Value = null);

public static class Validate
{
    // TODO[2020-11-01]: replace with FluentValidation or DataAnnotations
    private static readonly Regex EmailRe = new(@"^[^\s@]+@[^\s@]+\.[^\s@]+$", RegexOptions.Compiled);

    public static ValidationFailure Required(string field, string value) =>
        string.IsNullOrWhiteSpace(value)
            ? new ValidationFailure(field, "is required")
            : null;

    public static ValidationFailure Email(string field, string value) =>
        !EmailRe.IsMatch(value ?? "")
            ? new ValidationFailure(field, "must be a valid email address", value)
            : null;

    public static ValidationFailure MinLength(string field, string value, int min) =>
        (value?.Length ?? 0) < min
            ? new ValidationFailure(field, $"must be at least {min} characters", value)
            : null;

    public static ValidationFailure MaxLength(string field, string value, int max) =>
        (value?.Length ?? 0) > max
            ? new ValidationFailure(field, $"must be at most {max} characters", value)
            : null;

    public static IReadOnlyList<ValidationFailure> Collect(params ValidationFailure[] failures) =>
        failures.Where(f => f is not null).ToList();
}

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

public record Config(
    string Host,
    int    Port,
    string DbConnectionString,
    string JwtSecret,
    int    JwtExpirySec,
    int    CacheTtlSec,
    int    RateMaxRequests,
    int    RateWindowSec,
    string Environment)
{
    public static Config FromEnvironment() => new(
        Host:               Environment.GetEnvironmentVariable("HOST")          ?? "0.0.0.0",
        Port:               int.Parse(Environment.GetEnvironmentVariable("PORT") ?? "3000"),
        DbConnectionString: Environment.GetEnvironmentVariable("DB_URL")        ?? "Host=localhost;Database=app",
        JwtSecret:          Environment.GetEnvironmentVariable("JWT_SECRET")    ?? "change-me",
        JwtExpirySec:       int.Parse(Environment.GetEnvironmentVariable("JWT_EXPIRY") ?? "3600"),
        CacheTtlSec:        int.Parse(Environment.GetEnvironmentVariable("CACHE_TTL")  ?? "300"),
        RateMaxRequests:    int.Parse(Environment.GetEnvironmentVariable("RATE_MAX")   ?? "100"),
        RateWindowSec:      int.Parse(Environment.GetEnvironmentVariable("RATE_WINDOW") ?? "60"),
        Environment:        Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") ?? "Development");

    public bool IsProduction => Environment == "Production";
}

// ---------------------------------------------------------------------------
// Cache
// ---------------------------------------------------------------------------

// HACK[2019-06-01]: MemoryCache wrapper; replace with IDistributedCache + Redis
public sealed class Cache<TKey, TValue> : IDisposable where TKey : notnull
{
    private record Entry(TValue Value, DateTimeOffset ExpiresAt);

    private readonly ConcurrentDictionary<TKey, Entry> _store = new();
    private readonly Timer _cleanupTimer;

    public Cache()
    {
        _cleanupTimer = new Timer(_ => Cleanup(), null, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1));
    }

    public TValue Get(TKey key)
    {
        if (_store.TryGetValue(key, out var entry) && entry.ExpiresAt > DateTimeOffset.UtcNow)
            return entry.Value;
        _store.TryRemove(key, out _);
        return default;
    }

    public void Set(TKey key, TValue value, TimeSpan ttl) =>
        _store[key] = new Entry(value, DateTimeOffset.UtcNow.Add(ttl));

    public void Delete(TKey key) => _store.TryRemove(key, out _);

    public TValue GetOrSet(TKey key, TimeSpan ttl, Func<TValue> factory)
    {
        var cached = Get(key);
        if (cached is not null) return cached;
        var value = factory();
        Set(key, value, ttl);
        return value;
    }

    private void Cleanup()
    {
        var now = DateTimeOffset.UtcNow;
        foreach (var kvp in _store.Where(k => k.Value.ExpiresAt <= now))
            _store.TryRemove(kvp.Key, out _);
    }

    public void Dispose() => _cleanupTimer.Dispose();
}

// ---------------------------------------------------------------------------
// Rate limiter
// ---------------------------------------------------------------------------

public sealed class RateLimiter
{
    private record Entry(int Count, DateTimeOffset ResetAt);
    private readonly ConcurrentDictionary<string, Entry> _store = new();
    private readonly int _windowSec, _maxRequests;

    public RateLimiter(int windowSec, int maxRequests)
    {
        _windowSec   = windowSec;
        _maxRequests = maxRequests;
    }

    public (bool Allowed, int Remaining, int RetryAfterSec) Check(string key)
    {
        var now = DateTimeOffset.UtcNow;
        var entry = _store.AddOrUpdate(key,
            _ => new Entry(1, now.AddSeconds(_windowSec)),
            (_, e) => e.ResetAt > now ? e with { Count = e.Count + 1 } : new Entry(1, now.AddSeconds(_windowSec)));

        var remaining   = Math.Max(0, _maxRequests - entry.Count);
        var allowed     = entry.Count <= _maxRequests;
        var retryAfter  = allowed ? 0 : (int)(entry.ResetAt - now).TotalSeconds;
        return (allowed, remaining, retryAfter);
    }
}

// ---------------------------------------------------------------------------
// Pagination
// ---------------------------------------------------------------------------

public record Page<T>(IReadOnlyList<T> Items, int Total, int PageNum, int PageSize)
{
    public bool HasNext => (PageNum - 1) * PageSize + Items.Count < Total;
    public bool HasPrev => PageNum > 1;
}

public static class Pagination
{
    public static Page<T> Paginate<T>(IEnumerable<T> source, int page, int size)
    {
        var list   = source.ToList();
        var items  = list.Skip((page - 1) * size).Take(size).ToList();
        return new Page<T>(items, list.Count, page, size);
    }
}

// ---------------------------------------------------------------------------
// Feature flags
// ---------------------------------------------------------------------------

// TODO[2099-08-01][platform]: replace with Microsoft.FeatureManagement
public sealed class FeatureFlagService
{
    private record FlagDef(bool Enabled, int RolloutPercent, HashSet<string> Allowlist);
    private readonly ConcurrentDictionary<string, FlagDef> _flags = new();

    public void Define(string name, bool enabled, int rollout = 100, IEnumerable<string> allowlist = null)
    {
        _flags[name] = new FlagDef(enabled, rollout, new HashSet<string>(allowlist ?? []));
    }

    public bool IsEnabled(string name, string userId = null)
    {
        if (!_flags.TryGetValue(name, out var flag) || !flag.Enabled) return false;
        if (userId is not null && flag.Allowlist.Contains(userId)) return true;
        if (flag.RolloutPercent >= 100) return true;
        if (flag.RolloutPercent <= 0)  return false;

        using var md5    = MD5.Create();
        var hashBytes    = md5.ComputeHash(Encoding.UTF8.GetBytes($"{name}:{userId ?? "anon"}"));
        var bucket       = Math.Abs(BitConverter.ToInt32(hashBytes, 0)) % 100;
        return bucket < flag.RolloutPercent;
    }
}

// ---------------------------------------------------------------------------
// Circuit breaker
// ---------------------------------------------------------------------------

// FIXME[2021-03-15]: no half-open success threshold; add counter before closing
public sealed class CircuitBreaker
{
    private enum State { Closed, Open, HalfOpen }

    private State _state = State.Closed;
    private int _failures;
    private DateTimeOffset _lastFailure;
    private readonly int _threshold;
    private readonly TimeSpan _resetTimeout;
    private readonly object _lock = new();

    public CircuitBreaker(int threshold, TimeSpan resetTimeout)
    {
        _threshold    = threshold;
        _resetTimeout = resetTimeout;
    }

    public async Task<T> CallAsync<T>(Func<Task<T>> fn)
    {
        lock (_lock)
        {
            if (_state == State.Open)
            {
                if (DateTimeOffset.UtcNow - _lastFailure >= _resetTimeout)
                    _state = State.HalfOpen;
                else
                    throw new AppException("CIRCUIT_OPEN", "Service temporarily unavailable", 503);
            }
        }

        try
        {
            var result = await fn();
            lock (_lock) { _failures = 0; _state = State.Closed; }
            return result;
        }
        catch
        {
            lock (_lock)
            {
                _failures++;
                _lastFailure = DateTimeOffset.UtcNow;
                if (_failures >= _threshold) _state = State.Open;
            }
            throw;
        }
    }

    public string State => _state.ToString();
}

// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------

public static class StringUtils
{
    private static readonly Regex SlugRe = new(@"[^a-z0-9]+", RegexOptions.Compiled);

    public static string Slugify(string text) =>
        SlugRe.Replace(text.ToLowerInvariant(), "-").Trim('-');

    public static string MaskEmail(string email)
    {
        var parts = email.Split('@');
        if (parts.Length != 2) return email;
        var local   = parts[0];
        var visible = local.Length > 2 ? local[..2] : local[..1];
        var stars   = new string('*', Math.Max(1, local.Length - 2));
        return $"{visible}{stars}@{parts[1]}";
    }

    public static string Truncate(string s, int maxLen, string suffix = "…") =>
        s.Length <= maxLen ? s : s[..(maxLen - suffix.Length)] + suffix;
}

public static class FormatUtils
{
    public static string Bytes(long bytes)
    {
        string[] units = ["B", "KB", "MB", "GB", "TB"];
        double v = bytes;
        int i = 0;
        while (v >= 1024 && i < units.Length - 1) { v /= 1024; i++; }
        return $"{v:F2} {units[i]}";
    }

    public static string Duration(long ms) => ms switch
    {
        < 1000        => $"{ms}ms",
        < 60_000      => $"{ms / 1000.0:F1}s",
        _             => $"{ms / 60_000}m {ms % 60_000 / 1000}s"
    };
}

// TODO[2025-06-10]: add health-check endpoint wired to DI health checks pipeline
public sealed class HealthMonitor
{
    private readonly Dictionary<string, Func<Task<bool>>> _checks = new();
    private readonly DateTimeOffset _startedAt = DateTimeOffset.UtcNow;

    public void Register(string name, Func<Task<bool>> check) => _checks[name] = check;

    public async Task<object> RunAsync()
    {
        var results = new Dictionary<string, object>();
        foreach (var (name, check) in _checks)
        {
            var start = DateTimeOffset.UtcNow;
            try
            {
                var ok = await check();
                results[name] = new { status = ok ? "ok" : "fail", latencyMs = (DateTimeOffset.UtcNow - start).TotalMilliseconds };
            }
            catch (Exception ex)
            {
                results[name] = new { status = "fail", message = ex.Message };
            }
        }
        var healthy = results.Values.Cast<dynamic>().All(r => r.status == "ok");
        return new { status = healthy ? "healthy" : "unhealthy", uptime = (long)(DateTimeOffset.UtcNow - _startedAt).TotalSeconds, checks = results };
    }
}

File: tests/fixtures/sample.fs
// sample.fs — fixture file for timebomb scanner tests.
//
// Annotation inventory (hardcoded dates, never relative to today):
//   Expired        (2018–2021): 4
//   Expiring-soon  (2025-06):   1
//   Future / OK    (2088/2099): 2

module Sample

open System
open System.Collections.Concurrent
open System.Text.RegularExpressions
open System.Threading

// ---------------------------------------------------------------------------
// Domain types
// ---------------------------------------------------------------------------

type Environment = Development | Staging | Production

type Config =
    { Host           : string
      Port           : int
      DbUrl          : string
      JwtSecret      : string
      JwtExpiry      : TimeSpan
      CacheTtl       : TimeSpan
      RateMaxReqs    : int
      RateWindowSec  : int
      Env            : Environment }

let defaultConfig =
    { Host          = "0.0.0.0"
      Port          = 3000
      DbUrl         = "postgres://localhost/app"
      JwtSecret     = "change-me"
      JwtExpiry     = TimeSpan.FromHours 1.0
      CacheTtl      = TimeSpan.FromMinutes 5.0
      RateMaxReqs   = 100
      RateWindowSec = 60
      Env           = Development }

let isProduction cfg = cfg.Env = Production

// ---------------------------------------------------------------------------
// Result / validation
// ---------------------------------------------------------------------------

type ValidationFailure = { Field: string; Message: string; Value: string option }

type Validated<'a> =
    | Valid   of 'a
    | Invalid of ValidationFailure list

let mapValidated f = function
    | Valid a     -> Valid (f a)
    | Invalid errs -> Invalid errs

let bindValidated f = function
    | Valid a      -> f a
    | Invalid errs -> Invalid errs

let applyValidated fv av =
    match fv, av with
    | Valid f,   Valid a    -> Valid (f a)
    | Invalid e, Valid _    -> Invalid e
    | Valid _,   Invalid e  -> Invalid e
    | Invalid e1, Invalid e2 -> Invalid (e1 @ e2)

let required field (value: string) =
    if String.IsNullOrWhiteSpace value then
        Invalid [{ Field = field; Message = "is required"; Value = None }]
    else Valid value

// TODO[2020-09-01]: replace with a proper RFC 5322 email validator
let validateEmail (value: string) =
    let re = Regex(@"^[^\s@]+@[^\s@]+\.[^\s@]+$")
    if re.IsMatch(value) then Valid value
    else Invalid [{ Field = "email"; Message = "must be a valid email address"; Value = Some value }]

let minLength field min (value: string) =
    if value.Length >= min then Valid value
    else Invalid [{ Field = field; Message = $"must be at least {min} characters"; Value = Some value }]

let maxLength field max (value: string) =
    if value.Length <= max then Valid value
    else Invalid [{ Field = field; Message = $"must be at most {max} characters"; Value = Some value }]

// ---------------------------------------------------------------------------
// In-process cache
// ---------------------------------------------------------------------------

// HACK[2019-04-15]: ConcurrentDictionary cache; replace with StackExchange.Redis
type CacheEntry<'v> = { Value: 'v; ExpiresAt: DateTimeOffset }

type Cache<'k, 'v when 'k : equality>() =
    let store = ConcurrentDictionary<'k, CacheEntry<'v>>()

    member _.Get(key: 'k) =
        match store.TryGetValue(key) with
        | true, entry when entry.ExpiresAt > DateTimeOffset.UtcNow -> Some entry.Value
        | true, _ -> store.TryRemove(key) |> ignore; None
        | _ -> None

    member _.Set(key: 'k, value: 'v, ttl: TimeSpan) =
        store[key] <- { Value = value; ExpiresAt = DateTimeOffset.UtcNow.Add(ttl) }

    member _.Delete(key: 'k) =
        store.TryRemove(key) |> ignore

    member this.GetOrSet(key: 'k, ttl: TimeSpan, fn: unit -> 'v) =
        match this.Get(key) with
        | Some v -> v
        | None   ->
            let v = fn()
            this.Set(key, v, ttl)
            v

    member _.Cleanup() =
        let now = DateTimeOffset.UtcNow
        for kvp in store do
            if kvp.Value.ExpiresAt <= now then
                store.TryRemove(kvp.Key) |> ignore

// ---------------------------------------------------------------------------
// Rate limiter
// ---------------------------------------------------------------------------

type RateLimitResult = { Allowed: bool; Remaining: int; RetryAfterSec: int }

type RateLimiter(windowSec: int, maxRequests: int) =
    let store = ConcurrentDictionary<string, struct(int * DateTimeOffset)>()

    member _.Check(key: string) =
        let now = DateTimeOffset.UtcNow
        let struct(count, resetAt) =
            store.AddOrUpdate(
                key,
                struct(0, now.AddSeconds(float windowSec)),
                fun _ struct(c, exp) ->
                    if exp > now then struct(c, exp)
                    else struct(0, now.AddSeconds(float windowSec)))
        let newCount = count + 1
        store[key] <- struct(newCount, resetAt)
        let remaining   = max 0 (maxRequests - newCount)
        let allowed     = newCount <= maxRequests
        let retryAfter  = if allowed then 0 else int (resetAt - now).TotalSeconds
        { Allowed = allowed; Remaining = remaining; RetryAfterSec = retryAfter }

// ---------------------------------------------------------------------------
// Pagination
// ---------------------------------------------------------------------------

type Page<'a> =
    { Items   : 'a list
      Total   : int
      PageNum : int
      PageSize: int
      HasNext : bool
      HasPrev : bool }

let paginate (items: 'a list) pageNum pageSize =
    let offset = (pageNum - 1) * pageSize
    let chunk  = items |> List.skip (min offset items.Length) |> List.truncate pageSize
    { Items    = chunk
      Total    = items.Length
      PageNum  = pageNum
      PageSize = pageSize
      HasNext  = offset + pageSize < items.Length
      HasPrev  = pageNum > 1 }

// ---------------------------------------------------------------------------
// LRU cache (simple, not thread-safe)
// ---------------------------------------------------------------------------

// FIXME[2020-12-01]: LRU eviction is O(n) due to list scan; use a doubly-linked map
type LruCache<'k, 'v when 'k : equality>(capacity: int) =
    let mutable order : 'k list = []
    let store = Collections.Generic.Dictionary<'k, 'v>()

    member _.Get(key: 'k) =
        if store.ContainsKey(key) then
            order <- key :: List.filter ((<>) key) order
            Some store[key]
        else None

    member this.Set(key: 'k, value: 'v) =
        if store.ContainsKey(key) then
            order <- key :: List.filter ((<>) key) order
        else
            if store.Count >= capacity then
                let oldest = List.last order
                store.Remove(oldest) |> ignore
                order <- order |> List.filter ((<>) oldest)
            order <- [key] @ order
        store[key] <- value

    member _.Size = store.Count

// ---------------------------------------------------------------------------
// Feature flags
// ---------------------------------------------------------------------------

// TODO[2099-05-01][platform]: integrate with Azure App Configuration
type FlagDefinition =
    { Enabled        : bool
      RolloutPercent : int
      Allowlist      : Set<string> }

type FlagService() =
    let flags = ConcurrentDictionary<string, FlagDefinition>()

    member _.Define(name, enabled, rollout, ?allowlist) =
        flags[name] <- { Enabled = enabled; RolloutPercent = rollout
                         Allowlist = Set.ofList (defaultArg allowlist []) }

    member _.IsEnabled(name, ?userId) =
        match flags.TryGetValue(name) with
        | false, _ -> false
        | true, f  ->
            f.Enabled &&
            (match userId with
             | Some uid when f.Allowlist.Contains(uid) -> true
             | _ -> f.RolloutPercent >= 100)

// ---------------------------------------------------------------------------
// Event bus
// ---------------------------------------------------------------------------

type DomainEvent =
    { Id         : Guid
      Type       : string
      OccurredAt : DateTimeOffset
      Payload    : obj }

type EventHandler = DomainEvent -> unit

// FIXME[2025-06-08]: handlers run synchronously; move to async Task pipeline
type EventBus() =
    let handlers = ConcurrentDictionary<string, ResizeArray<EventHandler>>()

    member _.Subscribe(eventType: string, handler: EventHandler) =
        let list = handlers.GetOrAdd(eventType, fun _ -> ResizeArray())
        lock list (fun () -> list.Add(handler))
        fun () -> lock list (fun () -> list.Remove(handler) |> ignore)

    member _.Publish(eventType: string, payload: obj) =
        let evt = { Id = Guid.NewGuid(); Type = eventType
                    OccurredAt = DateTimeOffset.UtcNow; Payload = payload }
        match handlers.TryGetValue(eventType) with
        | true, list -> lock list (fun () -> list |> Seq.toArray) |> Array.iter (fun h -> h evt)
        | _          -> ()

// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------

let slugify (text: string) =
    text.ToLower()
    |> Seq.map (fun c -> if Char.IsLetterOrDigit c then c else '-')
    |> Seq.toArray
    |> String
    |> fun s -> Regex("-+").Replace(s, "-").Trim('-')

let maskEmail (email: string) =
    match email.Split('@') with
    | [| local; domain |] ->
        let visible = if local.Length > 2 then local.[..1] else local.[..0]
        let stars   = String.replicate (max 1 (local.Length - 2)) "*"
        $"{visible}{stars}@{domain}"
    | _ -> email

let truncate (maxLen: int) (s: string) =
    if s.Length <= maxLen then s
    else s.[..maxLen - 2] + "…"

let chunk (size: int) (lst: 'a list) =
    let rec go acc = function
        | [] -> List.rev acc
        | xs -> go (List.truncate size xs :: acc) (List.skip (min size xs.Length) xs)
    go [] lst

let groupBy (keyFn: 'a -> 'k) (lst: 'a list) =
    lst |> List.groupBy keyFn |> Map.ofList

// TODO[2088-12-01][observability]: expose metrics over HTTP /healthz
let formatBytes (bytes: int64) =
    let units = [| "B"; "KB"; "MB"; "GB"; "TB" |]
    let mutable i = 0
    let mutable v = float bytes
    while v >= 1024.0 && i < units.Length - 1 do
        v <- v / 1024.0
        i <- i + 1
    $"{v:F2} {units[i]}"

let formatDuration (ms: int64) =
    if ms < 1000L then $"{ms}ms"
    elif ms < 60_000L then $"{float ms / 1000.0:F1}s"
    else $"{ms / 60_000L}m {(ms % 60_000L) / 1000L}s"
